diff --git a/src/library_glfw.js b/src/library_glfw.js
index b71bab38e..05012d2e9 100644
--- a/src/library_glfw.js
+++ b/src/library_glfw.js
@@ -13,11 +13,9 @@
  * What it does not but should probably do:
  * - Transmit events when glfwPollEvents, glfwWaitEvents or glfwSwapBuffers is
  *    called. Events callbacks are called as soon as event are received.
- * - Joystick support.
  * - Input modes.
  * - Gamma ramps.
  * - Video modes.
- * - Monitors.
  * - Clipboard (not possible from javascript?).
  * - Multiple windows.
  * - Error codes && messages through callback.
@@ -71,6 +69,7 @@ var LibraryGLFW = {
       this.cursorPosFunc = null; // GLFWcursorposfun
       this.cursorEnterFunc = null; // GLFWcursorenterfun
       this.scrollFunc = null; // GLFWscrollfun
+      this.dropFunc = null; // GLFWdropfun
       this.keyFunc = null; // GLFWkeyfun
       this.charFunc = null; // GLFWcharfun
       this.userptr = null;
@@ -81,12 +80,14 @@ var LibraryGLFW = {
       return GLFW.windows[id - 1];
     },
 
+    joystickFunc: null, // GLFWjoystickfun
     errorFunc: null, // GLFWerrorfun
     monitorFunc: null, // GLFWmonitorfun
     active: null, // active window
     windows: null,
     monitors: null,
     monitorString: null,
+    videoMode: null,
     versionString: null,
     initialTime: null,
     extensions: null,
@@ -382,6 +383,14 @@ var LibraryGLFW = {
 #endif
     },
 
+    onGamepadConnected: function(event) {
+      GLFW.refreshJoysticks();
+    },
+
+    onGamepadDisconnected: function(event) {
+      GLFW.refreshJoysticks();
+    },
+
     onKeydown: function(event) {
       GLFW.onKeyChanged(event.keyCode, 1); // GLFW_PRESS or GLFW_REPEAT
 
@@ -641,6 +650,68 @@ var LibraryGLFW = {
       }
     },
 
+    setJoystickCallback: function(cbfun) {
+      GLFW.joystickFunc = cbfun;
+      GLFW.refreshJoysticks();
+    },
+
+    joys: {}, // glfw joystick data
+    lastGamepadState: null,
+    lastGamepadStateFrame: null, // The integer value of Browser.mainLoop.currentFrameNumber of when the last gamepad state was produced.
+
+    refreshJoysticks: function() {
+      // Produce a new Gamepad API sample if we are ticking a new game frame, or if not using emscripten_set_main_loop() at all to drive animation.
+      if (Browser.mainLoop.currentFrameNumber !== GLFW.lastGamepadStateFrame || !Browser.mainLoop.currentFrameNumber) {
+        GLFW.lastGamepadState = navigator.getGamepads ? navigator.getGamepads() : (navigator.webkitGetGamepads ? navigator.webkitGetGamepads : null);
+        GLFW.lastGamepadStateFrame = Browser.mainLoop.currentFrameNumber;
+
+        for (var joy = 0; joy < GLFW.lastGamepadState.length; ++joy) {
+          var gamepad = GLFW.lastGamepadState[joy];
+
+          if (gamepad) {
+            if (!GLFW.joys[joy]) {
+              console.log('glfw joystick connected:',joy);
+              GLFW.joys[joy] = {
+                id: allocate(intArrayFromString(gamepad.id), 'i8', ALLOC_NORMAL),
+                buttonsCount: gamepad.buttons.length,
+                axesCount: gamepad.axes.length,
+                buttons: allocate(new Array(gamepad.buttons.length), 'i8', ALLOC_NORMAL),
+                axes: allocate(new Array(gamepad.axes.length*4), 'float', ALLOC_NORMAL)
+              };
+
+              if (GLFW.joystickFunc) {
+                Module['dynCall_vii'](GLFW.joystickFunc, joy, 0x00040001); // GLFW_CONNECTED
+              }
+            }
+
+            var data = GLFW.joys[joy];
+
+            for (var i = 0; i < gamepad.buttons.length;  ++i) {
+              setValue(data.buttons + i, gamepad.buttons[i].pressed, 'i8');
+            }
+
+            for (var i = 0; i < gamepad.axes.length; ++i) {
+              setValue(data.axes + i*4, gamepad.axes[i], 'float');
+            }
+          } else {
+            if (GLFW.joys[joy]) {
+              console.log('glfw joystick disconnected',joy);
+
+              if (GLFW.joystickFunc) {
+                Module['dynCall_vii'](GLFW.joystickFunc, joy, 0x00040002); // GLFW_DISCONNECTED
+              }
+
+              _free(GLFW.joys[joy].id);
+              _free(GLFW.joys[joy].buttons);
+              _free(GLFW.joys[joy].axes);
+
+              delete GLFW.joys[joy];
+            }
+          }
+        }
+      }
+    },
+
     setKeyCallback: function(winid, cbfun) {
       var win = GLFW.WindowFromId(winid);
       if (!win) return;
@@ -671,6 +742,69 @@ var LibraryGLFW = {
       win.scrollFunc = cbfun;
     },
 
+    setDropCallback: function(winid, cbfun) {
+      var win = GLFW.WindowFromId(winid);
+      if (!win) return;
+      win.dropFunc = cbfun;
+    },
+
+    onDrop: function(event) {
+      if (!GLFW.active || !GLFW.active.dropFunc) return;
+      if (!event.dataTransfer || !event.dataTransfer.files || event.dataTransfer.files.length == 0) return;
+
+      event.preventDefault();
+
+      var filenames = allocate(new Array(event.dataTransfer.files.length*4), 'i8*', ALLOC_NORMAL);
+      var filenamesArray = [];
+      var count = event.dataTransfer.files.length;
+
+      // Read and save the files to emscripten's FS
+      var written = 0;
+      var drop_dir = '.glfw_dropped_files';
+      FS.createPath('/', drop_dir);
+
+      function save(file) {
+        var path = '/' + drop_dir + '/' + file.name.replace(/\//g, '_');
+        var reader = new FileReader();
+        reader.onloadend = function(e) {
+          if (reader.readyState != 2) { // not DONE
+            ++written;
+            console.log('failed to read dropped file: '+file.name+': '+reader.error);
+            return;
+          }
+
+          var data = e.target.result;
+          FS.writeFile(path, new Uint8Array(data), { encoding: 'binary' });
+          if (++written === count) {
+            Module['dynCall_viii'](GLFW.active.dropFunc, GLFW.active.id, count, filenames);
+
+            for (var i = 0; i < filenamesArray.length; ++i) {
+              _free(filenamesArray[i]);
+            }
+            _free(filenames);
+          }
+        };
+        reader.readAsArrayBuffer(file);
+
+        var filename = allocate(intArrayFromString(path), 'i8', ALLOC_NORMAL);
+        filenamesArray.push(filename);
+        setValue(filenames + i*4, filename, 'i8*');
+      }
+
+      for (var i = 0; i < count; ++i) {
+        save(event.dataTransfer.files[i]);
+      }
+
+      return false;
+    },
+
+    onDragover: function(event) {
+      if (!GLFW.active || !GLFW.active.dropFunc) return;
+
+      event.preventDefault();
+      return false;
+    },
+
     setWindowSizeCallback: function(winid, cbfun) {
       var win = GLFW.WindowFromId(winid);
       if (!win) return;
@@ -902,6 +1036,27 @@ var LibraryGLFW = {
       Module.ctx = Browser.destroyContext(Module['canvas'], true, true);
     },
 
+    getScreenVideoMode: function() {
+      if (!GLFW.videoMode) {
+        GLFW.videoMode = allocate([0, 0, 0, 0, 0, 0], 'i32', ALLOC_NORMAL);
+      }
+
+      setValue(GLFW.videoMode + 0*4, window.screen.width, 'i32');
+      setValue(GLFW.videoMode + 1*4, window.screen.height, 'i32');
+
+      var redBits = window.screen.colorDepth / 3;
+      var greenBits = window.screen.colorDepth / 3;
+      var blueBits = window.screen.colorDepth / 3;
+      setValue(GLFW.videoMode + 2*4, redBits, 'i32');
+      setValue(GLFW.videoMode + 3*4, greenBits, 'i32');
+      setValue(GLFW.videoMode + 4*4, blueBits, 'i32');
+
+      var refreshRate = 60; // TODO: estimate from requestAnimationFrame?
+      setValue(GLFW.videoMode + 5*4, refreshRate, 'i32');
+
+      return GLFW.videoMode;
+    },
+
     swapBuffers: function(winid) {
     },
 
@@ -954,6 +1109,8 @@ var LibraryGLFW = {
     GLFW.windows = new Array()
     GLFW.active = null;
 
+    window.addEventListener("gamepadconnected", GLFW.onGamepadConnected, true);
+    window.addEventListener("gamepaddisconnected", GLFW.onGamepadDisconnected, true);
     window.addEventListener("keydown", GLFW.onKeydown, true);
     window.addEventListener("keypress", GLFW.onKeyPress, true);
     window.addEventListener("keyup", GLFW.onKeyup, true);
@@ -965,6 +1122,8 @@ var LibraryGLFW = {
     Module["canvas"].addEventListener('mousewheel', GLFW.onMouseWheel, true);
     Module["canvas"].addEventListener('mouseenter', GLFW.onMouseenter, true);
     Module["canvas"].addEventListener('mouseleave', GLFW.onMouseleave, true);
+    Module["canvas"].addEventListener('drop', GLFW.onDrop, true);
+    Module["canvas"].addEventListener('dragover', GLFW.onDragover, true);
 
     Browser.resizeListeners.push(function(width, height) {
        GLFW.onCanvasResize(width, height);
@@ -973,6 +1132,8 @@ var LibraryGLFW = {
   },
 
   glfwTerminate: function() {
+    window.removeEventListener("gamepadconnected", GLFW.onGamepadConnected, true);
+    window.removeEventListener("gamepaddisconnected", GLFW.onGamepadDisconnected, true);
     window.removeEventListener("keydown", GLFW.onKeydown, true);
     window.removeEventListener("keypress", GLFW.onKeyPress, true);
     window.removeEventListener("keyup", GLFW.onKeyup, true);
@@ -984,6 +1145,10 @@ var LibraryGLFW = {
     Module["canvas"].removeEventListener('mousewheel', GLFW.onMouseWheel, true);
     Module["canvas"].removeEventListener('mouseenter', GLFW.onMouseenter, true);
     Module["canvas"].removeEventListener('mouseleave', GLFW.onMouseleave, true);
+    Module["canvas"].removeEventListener('drop', GLFW.onDrop, true);
+    Module["canvas"].removeEventListener('dragover', GLFW.onDragover, true);
+
+
     Module["canvas"].width = Module["canvas"].height = 1;
     GLFW.windows = null;
     GLFW.active = null;
@@ -1012,7 +1177,7 @@ var LibraryGLFW = {
   },
 
   glfwSetTime: function(time) {
-    GLFW.initialTime = GLFW.getTime() + time;
+    GLFW.initialTime = GLFW.getTime() - time;
   },
 
   glfwExtensionSupported: function(extension) {
@@ -1092,14 +1257,15 @@ var LibraryGLFW = {
     GLFW.monitorFunc = cbfun;
   },
 
-  // TODO: implement
   glfwGetVideoModes: function(monitor, count) {
-    setValue(count, 0, 'i32');
-    return 0;
+    setValue(count, 1, 'i32');
+
+    return GLFW.getScreenVideoMode();
   },
 
-  // TODO: implement
-  glfwGetVideoMode: function(monitor) { return 0; },
+  glfwGetVideoMode: function(monitor) {
+    return GLFW.getScreenVideoMode();
+  },
 
   // TODO: implement
   glfwSetGamma: function(monitor, gamma) { },
@@ -1189,6 +1355,8 @@ var LibraryGLFW = {
   glfwGetWindowMonitor: function(winid) {
     var win = GLFW.WindowFromId(winid);
     if (!win) return 0;
+    if (!GLFW.active.fullscreen) return 0;
+
     return win.monitor;
   },
 
@@ -1252,7 +1420,28 @@ var LibraryGLFW = {
 
   glfwFocusWindow: function(winid) {},
 
-  glfwSetWindowMonitor: function(winid, monitor, xpos, ypos, width, height, refreshRate) { throw "glfwSetWindowMonitor not implemented."; },
+  glfwSetWindowMonitor: function(winid, monitor, xpos, ypos, width, height, refreshRate) {
+    GLFW.active.monitor = monitor; // TODO: mask on whether actually in fullscreen
+    if (monitor) {
+      // TODO: use emscripten_request_fullscreen_strategy()
+      Module.requestFullscreen(1, 1);
+    } else {
+      // Null monitor exits fullscreen
+      // TODO: resize and reposition to given dimensions
+      // TODO: use emscripten_exit_fullscreen()
+      if (document.exitFullscreen) {
+        document.exitFullscreen();
+      } else if (document.msExitFullscreen) {
+        document.msExitFullscreen();
+      } else if (document.mozCancelFullScreen) {
+        document.mozCancelFullScreen();
+      } else if (document.webkitExitFullscreen) {
+        document.webkitExitFullscreen();
+      } else {
+        return {{{ cDefine('EMSCRIPTEN_RESULT_NOT_SUPPORTED') }}};
+      }
+    }
+  },
 
   glfwCreateCursor: function(image, xhot, yhot) {},
 
@@ -1340,7 +1529,9 @@ var LibraryGLFW = {
     return 0;
   },
 
-  glfwSetDropCallback: function(winid, cbfun) { throw "glfwSetDropCallback is not implemented."; },
+  glfwSetDropCallback: function(winid, cbfun) {
+    GLFW.setDropCallback(winid, cbfun);
+  },
 
   glfwGetTimerValue: function() { throw "glfwGetTimerValue is not implemented."; },
 
@@ -1354,15 +1545,49 @@ var LibraryGLFW = {
 
   glfwCreateWindowSurface: function(instance, winid, allocator, surface) { throw "glfwCreateWindowSurface is not implemented."; },
 
-  glfwSetJoystickCallback: function(cbfun) { throw "glfwSetJoystickCallback is not implemented."; },
+  glfwJoystickPresent: function(joy) {
+    GLFW.refreshJoysticks();
 
-  glfwJoystickPresent: function(joy) { throw "glfwJoystickPresent is not implemented."; },
+    return GLFW.joys[joy] !== undefined;
+  },
 
-  glfwGetJoystickAxes: function(joy, count) { throw "glfwGetJoystickAxes is not implemented."; },
+  glfwGetJoystickAxes: function(joy, count) {
+    GLFW.refreshJoysticks();
 
-  glfwGetJoystickButtons: function(joy, count) { throw "glfwGetJoystickButtons is not implemented."; },
+    var state = GLFW.joys[joy];
+    if (!state || !state.axes) {
+      setValue(count, 0, 'i32');
+      return;
+    }
+
+    setValue(count, state.axesCount, 'i32');
+    return state.axes;
+  },
 
-  glfwGetJoystickName: function(joy) { throw "glfwGetJoystickName is not implemented."; },
+  glfwGetJoystickButtons: function(joy, count) {
+    GLFW.refreshJoysticks();
+
+    var state = GLFW.joys[joy];
+    if (!state || !state.buttons) {
+      setValue(count, 0, 'i32');
+      return;
+    }
+
+    setValue(count, state.buttonsCount, 'i32');
+    return state.buttons;
+  },
+
+  glfwGetJoystickName: function(joy) {
+    if (GLFW.joys[joy]) {
+      return GLFW.joys[joy].id;
+    } else {
+      return 0;
+    }
+  },
+
+  glfwSetJoystickCallback: function(cbfun) {
+    GLFW.setJoystickCallback(cbfun);
+  },
 
   glfwSetClipboardString: function(win, string) {},
 
